Include command output when sending timeout to model

pakrym-oai opened the pull request:
Being able to see the output helps the model decide how to handle the timeout.

pakrym-oai committed:
Include command output when sending timeout to model

pakrym-oai committed:
core: refactor SandboxErr to use structured ExecToolCallOutput for Denied and Timeout variants

- Changed SandboxErr::Denied and SandboxErr::Timeout to carry boxed ExecToolCallOutput.
- Updated handling and matching of these variants throughout exec, codex, and error modules.
- Improved error reporting by referencing structured output fields directly.

pakrym-oai commented on codex-rs/core/src/error.rs:
```diff
@@ -13,8 +14,11 @@ pub type Result<T> = std::result::Result<T, CodexErr>;
+        "sandbox denied exec error, exit code: {}, stdout: {}, stderr: {}",
+        .output.exit_code, .output.stdout.text, .output.stderr.text
+    )]
+    Denied { output: Box<ExecToolCallOutput> },
```
box because error struct size is getting large

pakrym-oai committed:
tests/landlock: update match for Denied variant to use struct field syntax

pakrym-oai committed:
core/exec.rs: suppress unused_mut warning for timed_out in process_exec_tool_call

pakrym-oai committed:
tests/landlock: fix test_timeout panic match for Sandbox(Timeout)

aibrahim-oai commented on codex-rs/core/tests/suite/exec_stream_events.rs:
```diff
@@ -170,3 +173,35 @@ async fn test_aggregated_output_interleaves_in_order() {
+    assert_eq!(output.exit_code, 124);
+    assert_eq!(output.stdout.text, "before\n");
+    assert!(output.stderr.text.is_empty());
+    assert_eq!(output.aggregated_output.text, "before\n");
```
where is `command timed out after {} milliseconds\n` in the putput?

aibrahim-oai commented on codex-rs/core/src/codex.rs:
```diff
@@ -3114,7 +3118,7 @@ fn take_last_bytes_at_char_boundary(s: &str, maxb: usize) -> &str {
 
 /// Exec output is a pre-serialized JSON payload
-fn format_exec_output(exec_output: &ExecToolCallOutput) -> String {
+fn format_exec_output(exec_output: &ExecToolCallOutput, prefix: Option<String>) -> String {
```
non blocking: I think we can deduce the type from exec_output so we don't have to take prefix. However, I think it's ok.

bolinfest replied:
Can you document what `prefix` represents?

It feels a little off to me that we return `String` rather than `ExecOutput` (since the latter carries more information), but I guess this method is private to this file so the current signature is more convenient.

I mention this because if it returned `ExecOutput`, it would be easier to explain how `prefix` is used.

pakrym-oai replied:
Removed prefix.

pakrym-oai replied:
Added timed_out to the result.

pakrym-oai replied:
Added a separate test for content formatting. I'll merge this PR and see if I can get a functional test today as well.

aibrahim-oai reviewed (approved):
LGTM. Let's have an integration test to test the whole behavior or at least make sure that the prefix and the timeout appears in the test.

bolinfest commented on codex-rs/core/src/error.rs:
```diff
@@ -28,7 +32,7 @@ pub enum SandboxErr {
     /// Command timed out
     #[error("command timed out")]
-    Timeout,
+    Timeout { output: Box<ExecToolCallOutput> },
```
Should the `error()` include fields from `output` like `Denied`?

bolinfest commented on codex-rs/linux-sandbox/tests/suite/landlock.rs:
```diff
@@ -158,8 +158,8 @@ async fn assert_network_blocked(cmd: &[&str]) {
         Ok(output) => (output.exit_code, output.stdout.text, output.stderr.text),
-        Err(CodexErr::Sandbox(SandboxErr::Denied(exit_code, stdout, stderr))) => {
-            (exit_code, stdout, stderr)
+        Err(CodexErr::Sandbox(SandboxErr::Denied { output })) => {
```
Should this `match` just evaluate to `ExecToolCallOutput`?

pakrym-oai replied:
Yes

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -2887,15 +2888,18 @@ async fn handle_sandbox_error(
 
-    // if the command timed out, we can simply return this failure to the model
-    if matches!(error, SandboxErr::Timeout) {
+    if let SandboxErr::Timeout { output } = &error {
```
Is the case where the command times out, but is not run in a sandbox (i.e., `--dangerously-bypass-approvals-and-sandbox`) also covered by this case even though it isn't run in a "sandbox"?

pakrym-oai commented on codex-rs/core/src/exec.rs:
```diff
@@ -123,41 +124,55 @@ pub async fn process_exec_tool_call(
         Ok(raw_output) => {
-            let stdout = raw_output.stdout.from_utf8_lossy();
-            let stderr = raw_output.stderr.from_utf8_lossy();
+            #[allow(unused_mut)]
```
It is because of the cfg, I think #[allow(unused_mut)] is a bit cleaner.

pakrym-oai committed:
tests/landlock: simplify output handling and improve debug logging in assert_network_blocked

pakrym-oai committed:
exec: add timed_out flag to ExecToolCallOutput and propagate status

Include a timed_out boolean in ExecToolCallOutput to indicate if a command timed out. Update output formatting to prepend a timeout message when timed_out is true, and simplify related function signatures. Add test coverage for this behavior. Fix related error construction and consumers.

pakrym-oai committed:
landlock/tests: fix panic message to print stdout/stderr values correctly

pakrym-oai committed:
landlock: fix variable usage in network block assertion

Use output.exit_code instead of exit_code to correctly check the command exit status and ensure proper sandbox breach detection.
